docker-run.sh

```diff
@@ -0,0 +1,88 @@
+#!/bin/bash
+#
+# Copyright (c) 2019 Red Hat, Inc.
+# This program and the accompanying materials are made
+# available under the terms of the Eclipse Public License 2.0
+# which is available at https://www.eclipse.org/legal/epl-2.0/
+#
+# SPDX-License-Identifier: EPL-2.0
+#
+# Contributors:
+#   Red Hat, Inc. - initial API and implementation
+
+set
+
+# git ROOT directory used to mount filesystem
+GIT_ROOT_DIRECTORY=$(git rev-parse --show-toplevel)
+GO_MODULE=$(grep -e 'module ' ${GIT_ROOT_DIRECTORY}/go.mod | sed -e 's/module //')
+WORKDIR="/home/user/go/src/${GO_MODULE}"
+# Container image
+IMAGE_NAME="che-incubator/devworkspace-build-prerequisistes"
+
+# Operator SDK
+OPERATOR_SDK_VERSION=v0.12.0
+
+init() {
+  BLUE='\033[1;34m'
+  GREEN='\033[0;32m'
+  RED='\033[0;31m'
+  NC='\033[0m'
+  BOLD='\033[1m'
+}
+
+check() {
+  if [ $# -eq 0 ]; then
+    printf "%bError: %bNo script provided. Command is $ docker-run.sh <script-to-run> [optional-arguments-of-script-to-run]\n" "${RED}" "${NC}"
+    exit 1
+  fi
+  echo "check $1"
+  if [ ! -f "$1" ]; then
+    printf "%bError: %bscript %b provided is not existing. Command is $ docker-run.sh <script-to-run> [optional-arguments-of-script-to-run]\n" "${RED}" "${NC}" "${1}"
+    exit 1
+  fi
+}
+
+# Build image
+build() {
+  printf "%bBuilding image %b${IMAGE_NAME}${NC}..." "${BOLD}" "${BLUE}"
+  if docker build --build-arg OPERATOR_SDK_VERSION=${OPERATOR_SDK_VERSION} -t ${IMAGE_NAME} > docker-build-log 2>&1 -<<EOF
+  FROM python:3-alpine
+  ARG OPERATOR_SDK_VERSION
+  ENV GOROOT /usr/lib/go
+ 
+  RUN apk add --no-cache --update curl bash jq go \
+  && pip3 install yq \
+  && pip3 install jsonpatch
+
+  RUN curl -JL https://github.com/operator-framework/operator-sdk/releases/download/${OPERATOR_SDK_VERSION}/operator-sdk-${OPERATOR_SDK_VERSION}-x86_64-linux-gnu -o /bin/operator-sdk && chmod a+x /bin/operator-sdk
+  RUN mkdir -p /home/user/go && chmod -R a+w /home/user
+  ENV HOME /home/user
+  ENV GOPATH /home/user/go
+  WORKDIR ${WORKDIR}
+EOF
```

Comment:
Does building a dockerimage in this way allow caching of layers?

Reply:
Yes, afaik. @benoitf Do you confirm ?

Reply:
At least from what I experience locally it's not rebuilding the image if I run the build script a second time.

Reply:
yes

```bash
docker build -t foo -<<EOF
FROM python:3-alpine
RUN apk add --no-cache --update bash jq go
RUN apk add --no-cache --update curl
EOF
Sending build context to Docker daemon  2.048kB
Step 1/3 : FROM python:3-alpine
 ---> a1cd5654cf3c
Step 2/3 : RUN apk add --no-cache --update bash jq go
 ---> Running in 008183dd2fc5
fetch http://dl-cdn.alpinelinux.org/alpine/v3.11/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.11/community/x86_64/APKINDEX.tar.gz
(1/15) Installing bash (5.0.11-r1)
Executing bash-5.0.11-r1.post-install
(2/15) Installing libgcc (9.2.0-r3)
(3/15) Installing libstdc++ (9.2.0-r3)
(4/15) Installing binutils (2.33.1-r0)
(5/15) Installing gmp (6.1.2-r1)
(6/15) Installing isl (0.18-r0)
(7/15) Installing libgomp (9.2.0-r3)
(8/15) Installing libatomic (9.2.0-r3)
(9/15) Installing mpfr4 (4.0.2-r1)
(10/15) Installing mpc1 (1.1.0-r1)
(11/15) Installing gcc (9.2.0-r3)
(12/15) Installing musl-dev (1.1.24-r0)
(13/15) Installing go (1.13.4-r1)
(14/15) Installing oniguruma (6.9.4-r0)
(15/15) Installing jq (1.6-r0)
Executing busybox-1.31.1-r9.trigger
OK: 428 MiB in 50 packages
Removing intermediate container 008183dd2fc5
 ---> a0553fff2d8d
Step 3/3 : RUN apk add --no-cache --update curl
 ---> Running in 41fa96d1fb49
fetch http://dl-cdn.alpinelinux.org/alpine/v3.11/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.11/community/x86_64/APKINDEX.tar.gz
(1/3) Installing nghttp2-libs (1.40.0-r0)
(2/3) Installing libcurl (7.67.0-r0)
(3/3) Installing curl (7.67.0-r0)
Executing busybox-1.31.1-r9.trigger
OK: 429 MiB in 53 packages
Removing intermediate container 41fa96d1fb49
 ---> 725ca1c61c87
Successfully built 725ca1c61c87
Successfully tagged foo:latest





docker build -t foo -<<EOF
FROM python:3-alpine
RUN apk add --no-cache --update bash jq go
RUN apk add --no-cache --update curl
EOF
Sending build context to Docker daemon  2.048kB
Step 1/3 : FROM python:3-alpine
 ---> a1cd5654cf3c
Step 2/3 : RUN apk add --no-cache --update bash jq go
 ---> Using cache
 ---> a0553fff2d8d
Step 3/3 : RUN apk add --no-cache --update curl
 ---> Using cache
 ---> 725ca1c61c87
Successfully built 725ca1c61c87
Successfully tagged foo:latest
```